core/commonMain/src/kotlinx/serialization/encoding/Encoding.kt

```diff
@@ -486,3 +486,39 @@ public inline fun Encoder.encodeStructure(descriptor: SerialDescriptor, block: C
         if (ex == null) composite.endStructure(descriptor)
     }
 }
+
+/**
+ * Begins a collection, encodes it using the given [block] and ends it.
+ */
+public inline fun Encoder.encodeCollection(
+    descriptor: SerialDescriptor,
+    collectionSize: Int,
+    block: CompositeEncoder.() -> Unit
+) {
+    val composite = beginCollection(descriptor, collectionSize)
+    var ex: Throwable? = null
+    try {
+        composite.block()
+    } catch (e: Throwable) {
+        ex = e
+        throw e
+    } finally {
+        // End structure only if there is no exception, otherwise it can be swallowed
+        if (ex == null) composite.endStructure(descriptor)
+    }
+}
+
+/**
+ * Begins a collection, calls [block] with each item and ends the collections.
+ */
+public inline fun <E> Encoder.encodeCollection(
```

Comment:
Please add a test or a usage of this signature

Reply:
Done, I added a test in the json module which isn't ideal but I couldn't really figure out how to write an effective test in the core module.